cmd/daemonset-check/main.go

```diff
@@ -21,1118 +6,241 @@ import (
-			return err
-		}
-		log.Println("Successfully reported success to Kuberhealthy servers")
-	}
-
-	return nil
-}
-
-// doChecks actually runs checking procedures
-func (dsc *Checker) doChecks() error {
-
-	// deploy the daemonset
-	err := dsc.doDeploy()
-	if err != nil {
-		return err
-	}
-
-	// clean up the daemonset.  Does not return until removed completely or
-	// an error occurs
-	err = dsc.doRemove()
-	if err != nil {
-		return err
-	}
-
-	// fire off an orphan cleanup in the background on each check run
-	go dsc.cleanupOrphans()
-
-	return nil
-}
-
-// waitForAllDaemonsetsToClear
-func (dsc *Checker) waitForAllDaemonsetsToClear() chan error {
-	log.Infoln("waiting for all daemonsets to clear")
-
-	// make the output channel we will return and close it whenever we are done
-	outChan := make(chan error, 2)
+	// make a new context for our shutdown timeout
+	shutdownCtx, shutdownCtxCancel := context.WithTimeout(context.Background(), shutdownGracePeriod)
+	defer shutdownCtxCancel()
 
+	// start a shutdown in the background
+	shutdownCompleteChan := make(chan error)
 	go func() {
-		// watch events and return when the pod is in state running
-		for {
-			log.Debugln("Waiting for daemonset", dsc.Name(), "to clear...")
-
-			// wait between requests
-			time.Sleep(time.Second * 5)
-
-			// if the context is canceled, we stop
-			select {
-			case <-dsc.ctx.Done():
-				outChan <- errors.New("waiting for daemonset to clear was aborted by context cancellation")
-				return
-			default:
-			}
-
-			// fetch the pod by name
-			dsList, err := dsc.getAllDaemonsets()
-			if err != nil {
-				outChan <- err
-				return
-			}
-			if len(dsList) == 0 {
-				log.Info("all daemonsets cleared")
-				outChan <- nil
-				return
-			}
-		}
+		shutdownCompleteChan <- shutdown(shutdownCtx)
 	}()
 
-	return outChan
-}
-
-// fetchDS fetches the ds for the checker from the api server
-// and returns a bool indicating if it exists or not
-func (dsc *Checker) fetchDS() (bool, error) {
-	dsClient := dsc.getDaemonSetClient()
-	var firstQuery bool
-	var more string
-	// pagination
-	for firstQuery || len(more) > 0 {
-		firstQuery = false
-		dsList, err := dsClient.List(metav1.ListOptions{
-			Continue: more,
-		})
+	// watch for timeout or shutdown to complete
+	select {
+	case err := <-shutdownCompleteChan:
 		if err != nil {
-			return false, err
-		}
-		more = dsList.Continue
-
-		// check results for our daemonset
-		for _, item := range dsList.Items {
-			if item.GetName() == dsc.DaemonSetName {
-				// ds does exist, return true
-				return true, nil
-			}
+			log.Infoln("shutdown completed with error:", err)
+			os.Exit(1)
 		}
+		log.Infoln("shutdown completed without error")
+		os.Exit(0)
+	case <-time.After(time.Duration(shutdownGracePeriod)):
+		log.Errorln("Shutdown took too long. Shutting down forcefully!")
+		os.Exit(2)
 	}
-
-	// daemonset does not exist, return false
-	return false, nil
+	os.Exit(0)
```

Comment:
I think you might want to defer the `os.Exit()` calls to allow `shutdownCtxCancel()` to be called before the process exits since `shutdownCtxCancel()` was deferred